### ISSUE TITLE
Tags attribute isn't really inheritable (even though it's "tagged" as such)...

### ISSUE THREAD

**Issue by zendever** [schambers/fluentmigrator#626]
This may have been a discussion somewhere along the way, or there may be some other way to do this, but setting the Tag attribute on a base migration class doesn't work as I would expect.

Case in point:

on DefaultMigrationConventionTests.cs:

```c#
// inherit from tagged class
public class InheritedFromTaggedWithUk : TaggedWithUk
{
}
...
// test along side tag tester...test will fail
[Test]
public void TypeHasTagsReturnTrueIfTypeHasTagsAttribute()
{
	DefaultMigrationConventions.TypeHasTags(typeof(TaggedWithUk))
		.ShouldBeTrue();

	DefaultMigrationConventions.TypeHasTags(typeof(InheritedFromTaggedWithUk))
		.ShouldBeTrue();
}
```

The code that tests for tag existence clearly only allows the Tags attribute to act as if they were non-inheritable:

in DefaultMigrationConventions.cs

```c#
public static bool TypeHasTags(Type type)
{
	return type.GetOneAttribute<TagsAttribute>() != null; // <-- note the overload with no param
}
```

in ExtensionsForICustomAttributeProvider.cs:

```c#
// the function below IS called for the Tags attribute
public static T GetOneAttribute<T>(this ICustomAttributeProvider member)
	where T : Attribute
{
	return member.GetOneAttribute<T>(false);
}

// the function below is NOT called for the Tags attribute
public static T GetOneAttribute<T>(this ICustomAttributeProvider member, bool inherit)
	where T : Attribute
{
	T[] attributes = member.GetCustomAttributes(typeof(T), inherit) as T[];

	if ((attributes == null) || (attributes.Length == 0))
		return null;
	else
		return attributes[0];
}
```

changing the DefaultMigrationConventions.cs class to the code below seems to pass all unit tests, including the Tag inheritance test above.

```c#
public static bool TypeHasTags(Type type)
{
	return type.GetOneAttribute<TagsAttribute>(true) != null;
}
```

Again, I realize that this may be one of those finer discussion points on the "proper" usage of migrations. Maybe we didn't intend for the Migration class to be turned into a base class. We could certainly argue about whether or not Tags should be "overridden" if they're inherited. We could certainly argue about composition over inheritance...that said, if any of those were really issues, the solution would simply be to make Tags not inheritable, right? At least that way, there'd be no confusion over their purpose.

No that I mind typing the Tags attribute for every migration...but that's a workaround. Real world use case is as simple as wanting to create a database with specific directions, for specific environments. Yes, we could absolutely create 2 migration projects, and then provide a build script or something...but again, Tags seemed a viable alternative. It just goes back to the reason why the Tags attribute was marked as Inherited = true in the first place.

Thoughts?

**Comment by zendever** [comment]
@tommarien ... Any thoughts as to this small change? I understand that explicitly tagging a migration may enforce better planning, but having the option to tag a set of migrations could also be useful. I know there may be other workarounds (like using namespaces), but that seems like coupling concerns.

I'll certainly take suggestions if we want to keep tags as forcibly "uninheritable", but it think it'd be a useful feature to have.

Again, I'm willing to make a pull request as long as the feature adheres to your plans.

**Comment by tommarien** [comment]
@zendever Can you explain to me what is for you difference with allowing the tag attribute to be inherited or the addition of multiple tags attributes

To be clear it's only possible when you create your own base migration which is tagged, so you will have to remember to inherit from it. Is this not exactly the same as having all migrations tagged ? I think the second (== current) is far more extendable and easier. For me it's like 'favor composition over inheritance" ;)  But on a code side it seems the same no ?

**Comment by zendever** [comment]
@tommarien ... for some of our enterprise migrations, we have multiple features that may need to go into multiple databases as part of a larger enterprise migration. So rather than having multiple migration projects, we use the migration architecture to implement enterprise features across several databases.

For example, we have a use case for standing up several service bus endpoints (which is less about migration and more about development) and consume 2 databases (1 for logging and 1 for the core data) for the entire bus. This is easier to maintain in 1 project rather than 3+. So, I'm using Tags to differentiate between the different connections.

When we implement features, sometimes they affect what we do to the logging database (for data types, events, etc) AND the core database. So we use multiple contexts that are migrated in the same project. This way, we can run 1 set of migrations to update the features for both databases. Again, tags let us do this.

Now, I use the Tags in combination with the MSBuild tasks (and namespaces...but for different reasons). This allows me to inherit from a base class (which inherits from Migration) and do some pre/post work depending on the target catalog.

Now, because of the way this is set up, we know that inheriting from the catalog migration base which catalog it's targeting. If we could tag the base class, then I wouldn't have to tag each individual migration...it's really just saving keystrokes. You can see, more clearly, what I mean by the code examples I have in the original post.

As I pointed out, if the Tags attribute wasn't mean to be inherited such that tagging a base class would also attribute any successors, then the TagsAttribute class should have set the Inherited property to FALSE. This would allay any confusion. From looking at the TagsAttribute source, I would have just assumed I could attribute the base class, and then have all of its successors adhere to the Tag, but it simply failed, and I had to dig around the source to figure out why. I added the unit test above to show this failure and explained why it fails (it's coded to only select from the top most class' tags GetOneAttribute function).

Typically, I'd agree that composition is preferable over inheritance, but inheritance certainly has its place. What I'm suggesting doesn't remove the possibility of tag composition, it simply allows for the inheritance of them, as well.

**Comment by tommarien** [comment]
@zendever accepted ;) would have no mayor drawbacks anyway, just letting tags be inheritable

**Comment by zendever** [comment]
Okay, I'll code up the tests and create a PR

**Comment by tommarien** [comment]
@zendever Thanks and sorry for being a pita :+1:

**Comment by zendever** [comment]
@tommarien ... Not at all...First, this is your vision. Second, I'm certainly fallible. and Last, this is the purpose of GitHub (IMO), we discuss the best direction for the project to produce a better product. :-)

**Comment by tommarien** [comment]
@zendever agreed :+1:

**Comment by cbright** [comment]
zendever,

I ran into this exact issue and did what you did an pulled down the code and came to same conclusion. However, I think the TypeHasMatchingTags method also needs to be modified to pass TRUE to  GetAllAttributes. 

I think changing just TypeHasTags with out changing TypeHasMatchingTags seems to break a few of the integration tests if you have one of the include migrations inherit for a base migration and put the tag on the base. Such as

```c#
    [Tags("TenantA")]
    public abstract class TenantABaseMigration : Migration
    {
    }

    [Migration(1)]
    public class TenantATable : TenantABaseMigration
    {
        private const string TableName = "TenantATable";

        public override void Up()
        {
            Create.Table(TableName)
                .WithColumn("Id").AsInt32();
        }

        public override void Down()
        {
            Delete.Table(TableName);
        }
    }

        [Test]
        public void MigrateUpWithTaggedMigrationsShouldOnlyApplyMatchedMigrations()
        {
            ExecuteWithSupportedProcessors(processor =>
            {
                var assembly = typeof(TenantATable).Assembly;

                var runnerContext = new RunnerContext(new TextWriterAnnouncer(System.Console.Out))
                {
                    Namespace = typeof(TenantATable).Namespace,
                    Tags = new[] { "TenantA" }
                };

                var runner = new MigrationRunner(assembly, runnerContext, processor);

                try
                {
                    runner.MigrateUp(false);

                    processor.TableExists(null, "TenantATable").ShouldBeTrue();
                    processor.TableExists(null, "NormalTable").ShouldBeTrue();
                    processor.TableExists(null, "TenantBTable").ShouldBeFalse();
                    processor.TableExists(null, "TenantAandBTable").ShouldBeTrue();
                }
                finally
                {
                    runner.RollbackToVersion(0);
                }
            });
        }
```

The test fails on the "TenantATable" assertion because I don't think it can see the Tag still because of the following in DefaultMigrationInformationLoader

```C#
        private IEnumerable<Type> FindMigrationTypes()
        {
            return Assemblies.GetExportedTypes()
                .FilterByNamespace(Namespace, LoadNestedNamespaces)
                .Where(t => Conventions.TypeIsMigration(t)
                            &&
                            (Conventions.TypeHasMatchingTags(t, TagsToMatch) ||
                             !Conventions.TypeHasTags(t)));
        }
```

Anyway I built upon your solution with a few more tests https://github.com/cbright/fluentmigrator/commit/9dceb60f5b43cf031dccbdf55354f94e0a719d35. Let me know if I'm completely off base and this isn't really the issue you are having. I'm not trying to hijack the thread.

**Comment by tommarien** [comment]
@zendever and @cbright I totally agree, pulling @cbright pr seems like the right approach ;)

**Comment by zendever** [comment]
I also concur. Good catch!

**Comment by tommarien** [comment]
Closed by merging #648
